File: Iterator/Golang/UserIterator.go
package main

type UserIterator struct {
	Position int
	Collection []*User
}

func (userIterator *UserIterator) hasNext() bool  {
	return userIterator.Position < len(userIterator.Collection)
}

func (userIterator *UserIterator) next() *User  {
	if userIterator.Position < len(userIterator.Collection) {
		user := userIterator.Collection[userIterator.Position]

		if userIterator.hasNext() {
			userIterator.Position++
		}

		return user
	}

	return nil
}

File: Iterator/Dart/UserIterator.dart
import 'IIterator.dart';
import 'User.dart';
import 'UserCollection.dart';

class UserIterator implements IIterator {
  List<User>? _collection;
  int position = 0;

  UserIterator(this._collection);

  @override
  bool hasNext() {
    var length = _collection?.length ?? 0;

    return position < length;
  }

  @override
  next() {
    if (hasNext()) {
      return _collection![position++];
    }
  }
}
